### PR TITLE
Aggregates API

### PR DESCRIPTION
**Pull request by sileht**
Design is discussed here: https://github.com/gnocchixyz/gnocchi/issues/393

### CODE REVIEW

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 137f25318a30cd31d07dd1fd3d6a88e7543fcd32

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 48d8cc6c67bb96d972522fc60b6afefc6b8cde4d

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 4560861759225bcee28fc7902fdc4fa404e2779d

**Comment by sileht** [comment]
Pull-request updated, HEAD is now b5db8301cd751d424743de7f456fa8159879e66b

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 7094437314ba37b352b1e8a1f1bf307d86e22fc9

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 750028a241ca98472950c0c9e7a1d91611cf0558

**Comment by sileht** [comment]
Pull-request updated, HEAD is now fb22fa9d9528370001732fb80039bf139ae6d9dd

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 9ceee66d0a89024d502187d4248a6b65c40a90ce

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 26d81d4fb5faf972caf0e6875b1029a805a7d093

**Comment by sileht** [comment]
Pull-request updated, HEAD is now ce29b9456c0f17b29675fd5ab2f10521f95b2ba6

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 77fdaded64ca3f66a503cb1ff30dbf6a30b9468d

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 76c1137751f88b8bb97efc13fce261dbbe946678

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 9fde13a6bc8a0508b7da48b1147613d99fd0ffe6

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 870f89807a2356e44d025b865a4e5c9199bc18ff

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 173ed8222d575f051637ce544998a2bbaa7e6d4e

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 85a408c0a8dbe093b67fa7f6ba172cd4aafa8cbe

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 767fac154787f28b82f2dca3578609bdafa01035

**Comment by sileht** [comment]
Pull-request updated, HEAD is now b519090eabbda005ca5f5b238ccd3543800c3e8b

**Comment by sileht** [comment]
Pull-request updated, HEAD is now af50364a9e5d73067d4c3320665d7652bbadd1aa

File: gnocchi/tests/functional/gabbits/aggregates.yaml [hunk 1]
```diff
@@ -0,0 +1,212 @@
+fixtures:
```
Comment (jd) [inline gnocchi/tests/functional/gabbits/aggregates.yaml]: it'd be great to have one scenario for fetch and one for search-and-fetch

File: gnocchi/rest/aggregates/api.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            if fill != 'null':
+                api.abort(400, "fill must be a float or \'null\': %s" % e)
+    return start, stop, granularity, needed_overlap, fill
+
+
+def OperationsSchema(v, metrics_schema, root_node=False):
+    subnode = functools.partial(OperationsSchema,
+                                metrics_schema=metrics_schema)
+
+    operations_schema = [
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/rest/aggregates/api.py]: The whole schema uses `metric_schema` that is generated dynamically with values in `metric/metric_names`

File: gnocchi/tests/functional/gabbits/aggregates.yaml [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+          - timestamp: "2015-03-06T14:34:15"
+            value: -16
+          - timestamp: "2015-03-06T14:35:12"
+            value: 9
+          - timestamp: "2015-03-06T14:35:15"
+            value: 11
+      status: 202
+
+    - name: push measurements to metric2
+      POST: /v1/metric/$HISTORY['list valid metrics'].$RESPONSE['$[1].id']/measures
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/tests/functional/gabbits/aggregates.yaml]: done

File: gnocchi/tests/functional/gabbits/aggregates.yaml [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      data:
+          archive_policy_name: cookies
+      status: 201
+      response_json_paths:
+          $.archive_policy_name: cookies
+
+    - name: list valid metrics
+      GET: /v1/metric
+      response_json_paths:
+          $[0].archive_policy.name: cookies
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/tests/functional/gabbits/aggregates.yaml]: done

File: gnocchi/tests/functional/gabbits/aggregates.yaml [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+          $.name: disk.io.rate
+          $.unit: B/s
+
+    - name: create valid metric
+      POST: /v1/metric
+      data:
+          archive_policy_name: cookies
+      status: 201
+      response_json_paths:
+          $.archive_policy_name: cookies
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/tests/functional/gabbits/aggregates.yaml]: done

File: gnocchi/tests/functional/gabbits/aggregates.yaml [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+              - granularity: 5 second
+      status: 201
+
+    - name: create metric with name and unit
+      POST: /v1/metric
+      data:
+          name: "disk.io.rate"
+          unit: "B/s"
+      status: 201
+      response_json_paths:
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/tests/functional/gabbits/aggregates.yaml]: done

File: gnocchi/tests/functional/gabbits/aggregates.yaml [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        archive_policy_name: cookies
+      status: 201
+
+    - name: create alt archive policy
+      POST: /v1/archive_policy
+      request_headers:
+        # User admin
+        authorization: "basic YWRtaW46"
+      data:
+          name: cream
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/tests/functional/gabbits/aggregates.yaml]: done

File: gnocchi/tests/functional/gabbits/aggregates.yaml [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+      request_headers:
+        # User admin
+        authorization: "basic YWRtaW46"
+      data:
+          name: cookies
+          definition:
+              - granularity: 1 second
+      status: 201
+
+    - name: create archive policy rule
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/tests/functional/gabbits/aggregates.yaml]: done

File: gnocchi/tests/functional/gabbits/aggregates.yaml (hunk 1 above)
Comment (sileht) [reply gnocchi/tests/functional/gabbits/aggregates.yaml]: done

File: gnocchi/rest/aggregates/api.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+            api.abort(400, e)
+
+    if fill is not None:
+        if granularity is None:
+            api.abort(400, "Unable to fill without a granularity")
+        try:
+            fill = float(fill)
+        except ValueError as e:
+            if fill != 'null':
+                api.abort(400, "fill must be a float or \'null\': %s" % e)
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/rest/aggregates/api.py]: done

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 266222bdf7fde61e2ade5a19cedbe3d731ecdb4e

**Comment by sileht** [comment]
Pull-request updated, HEAD is now ce353f193a2027a6635acd187ec72b25c6c5d284

File: gnocchi/rest/aggregates/api.py [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+        body = api.deserialize_and_validate(self.FetchSchema)
+
+        metric_ids = [six.text_type(m[0]) for m in body["metrics"]]
+        metrics = pecan.request.indexer.list_metrics(ids=metric_ids)
+        missing_metric_ids = (set(metric_ids)
+                              - set(six.text_type(m.id) for m in metrics))
+        if missing_metric_ids:
+            # Return one of the missing one in the error
+            api.abort(404, storage.MetricDoesNotExist(
+                missing_metric_ids.pop()))
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/rest/aggregates/api.py]: done

File: gnocchi/rest/aggregates/api.py [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+        try:
+            return processor.get_measures(
+                pecan.request.storage,
+                metrics_and_aggregations,
+                operations,
+                start, stop,
+                granularity, needed_overlap, fill)
+        except processor.MetricUnaggregatable as e:
+            api.abort(400, ("One of the metrics being aggregated doesn't have "
+                            "matching granularity: %s") % str(e))
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/rest/aggregates/api.py]: done

File: gnocchi/rest/aggregates/api.py [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+            return processor.get_measures(
+                pecan.request.storage,
+                metrics_and_aggregations,
+                operations,
+                start, stop,
+                granularity, needed_overlap, fill,
+                ref_identifier="name")
+        except processor.MetricUnaggregatable as e:
+            api.abort(400, ("One of the metrics being aggregated doesn't have "
+                            "matching granularity: %s") % str(e))
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/rest/aggregates/api.py]: done

File: gnocchi/rest/aggregates/api.py [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+        except (storage.MetricDoesNotExist,
+                storage.GranularityDoesNotExist,
+                storage.AggregationDoesNotExist) as e:
+            api.abort(404, e)
```
[the remark this answers is not in the document]
Comment (sileht) [reply gnocchi/rest/aggregates/api.py]: done

**Comment by sileht** [comment]
Pull-request updated, HEAD is now ae5a60deb5926c924151b84a267985dc2f45eba4

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 146ca20c9744b65d0b7c0c10dfcdb57de92b3ecc

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 3848c441a065fc03f28c2e2a4aecc07e7be08491

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 9c52983256ea71397b3b3b49d40f56c101ca5474

**Comment by sileht** [comment]
Pull-request updated, HEAD is now e2eedeaabb7ac3eb8ccd6cf88c8e4179c375148f

**Comment by jd** [comment]
@sileht this pr is really big and has a lot of commits. I'm afraid it's gonna be hard to review and merge. Can you try to send a few smaller PR before? If there are things you can extract.

**Comment by sileht** [comment]
Pull-request updated, HEAD is now cca499c8e3d50f720b8feadb078a9d8f8ffe8fcb

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 243e8ee808c1b3ccd36abf02dad2bd30f58c110b

**Comment by sileht** [comment]
Everything have been splitted and submitted.

**Pull request closed without merging**